### ISSUE TITLE
Select2 loses focus when {closeOnSelect: false}

### ISSUE THREAD

**Issue by ArxSargos** [select2/select2#3275]
I tried to get through all issues with focus, but didn't found similar or duplicate issue. Sorry if this is solved or already tracked as issue...

When {closeOnSelect: false} is set, after selection made by keaboard select2 loses it's focus. http://jsfiddle.net/522bsupt/  (comparison of standard and closeOnSelect:false option)

**Comment by piotr-cz** [comment]
I must confirm the issue

**Comment by gryn** [comment]
I'd like to confirm that the behavior with closeOnSelect=false doesn't work at all really. I'm working on having select2 act more like a menu with checkboxes, so I am using SingleSelection as my working base for a custom Adapter on a <select multiple>. 

There are at least two issues:

1) Changing selection via keyboard or mouse steals the highlight/focus to be the first selected item. (this is due to the code in results.setClasses -- I have made a monkey patch to work around this issue).

2) Using the keyboard's ENTER will only select a new item (and then lose focus to the first selected item), using ENTER on an already select item *closes* the dropdown (which directly violates the closeOnSelect=false option). Using space+control is technically the correct key press, but with closeOnSelect=false I feel ENTER and space+control should work the same. However, space+control works but still fails to keep focus in the correct place.

In summary, the focus is completely messed up and ENTER should behave differently if closeOnSelect=false.

**Comment by gryn** [comment]
Here is my focus fix hack:
				select.on('select2:selecting', recordFocus);
				select.on('select2:unselecting', recordFocus);
				select.on('select2:select', setFocus);
				select.on('select2:unselect', setFocus);

				var curOption;

				function recordFocus(e) {
					var data = e.params.args.data;

					// select2 doesn't create jquery friendly id's (they may have funny chars)
					// so use getElementById
					curOption = data && data._resultId ?
						$(document.getElementById(data._resultId)) :
						null;
				}
				function setFocus(e) {
					curOption && curOption.trigger('mouseenter').trigger('focus');
				}

**Comment by gryn** [comment]
And I won't post the hack I have done for the keys, since there is no clean way to change the keyboard behavior (it's added globally at initialization and can't be customized or disabled). But effectively I was able to make everything "work" by changing the behavior for KEYS.ENTER to be results:toggle, and also I removed the control key requirement for the space bar (since that's how checkboxes would work and that's how my UI is presented in my case). Being able to customize this behavior would be nice instead of having to hack it. (Just allowing an override function is better than nothing).

**Comment by piotr-cz** [comment]
@gryn Please submit your hack if you can. It may save mew few hours

**Comment by gryn** [comment]
I basically overrode the existing keypress function with a copy. (I did this by doing something like: select2.listeners.keypress[0] = function(evt) { ... } ). The lines changed are just the KEYS.ENTER block calls the toggle event, and the KEYS.SPACE block doesn't require control to be pressed. As a simple customization compromise, it would be nice if this entire handler was optionally disabled. But something more sophsiticated would be fine too of course :) .


Concerning the other issue with close on select (results.setClasses reseting focus), you probably need select2 to actually remember the focus state (or ask the DOM), and fallback on the existing behavior if that fails. (this would be good also if you want the state of the dropdown to possibly be maintained between open and close). Anyway, good luck.

**Comment by piotr-cz** [comment]
@gryn 
Somehow you code wasn't working for me. It was somehow = when I flipped the event names (select <-> unselect) but sometimes the _resultId property was missing.

This is what I came up with:

````
// Get the Select2 instasnce
var select2 = this.$select.data('select2');

select2.on('select', function(event) {
    // Reopen the results dialog
    this.trigger('open');

    var $curOption;
    var $results = this.$results.find('.select2-results__option');

    // Find matching option element
    $results.each(function(i, $element) {
        if (jQuery($element).data('data') == event.data) {
            $curOption = jQuery($element);
        }
    });

    if ($curOption.length) {
        $curOption.trigger('mouseenter');
    }
});
```

**Comment by piotr-cz** [comment]
And this is my fix to use Enter key for unselecting items

```js
// Use enter for unselecting items in list, just like click
var select2 = this.$select.data('select2');

select2.on('keypress', function(evt) {
     // Funny but opposite works
     if (!this.isOpen() && evt.which === 13) {
          this.trigger('results:toggle', {originalEvent: evt});
    }
});
 ```

**Comment by gryn** [comment]
Piotr, I think it's reversed because you are adding your event on top of an existing (already processed) event, so the dropdown is already open by the time you get there (because by default enter selects the item then closes the drop down). Also for your reselecting, I would change the line var $curOption; to read var $curOption = $(); (or else change the if statement to avoid calling length on an undefined value).

**Comment by piotr-cz** [comment]
Thanks for tips.

By the way, the issue is still present in the v4.0.0

**Comment by abhilashpullelil** [comment]
Hello everyone,

I know it is too late to post here, but someone like me is still searching for a solution to this problem.

I faced the same issue and I have fixed the issue by updating the select2 library. But I have updated the older version of the select2 (Version: 3.5.1). So I can't send pull request to current version of select2. Pardon me!!!

If you are using version 3.5.1 then you can simply use my updated library.

**You can find this library here,**

https://github.com/abhilashpullelil/select2/blob/master/dist/js/select2.js

**Comment by inieves** [comment]
Id like to comment that I am able to get good keyboard focus handling from the fiddle at: http://jsfiddle.net/522bsupt/

However, as soon as you change the multiselects to single selects, by removing the multiple="" parts, the keyboard focus handling breaks.

I cannot tab into a select, select an item, and then tab to the next select.  I tested on latest Safari on Mac.

-ian

**Comment by RajendarKumar** [comment]
HI All,
How to keep focus on selected item in multi select

see below URL
https://select2.github.io/examples.html#hide-search

Go to 
Multiple select boxes

1.  Now select first time with pressing CTRL key
2.  Now select item after scrolling down at second page with CTRL key
3. Now focus will be shifted on first item

My Requirement is, focus should be on the same item, it must not shift onto fist item
I have tried with below code but it is not working

var $eventSelect = $(".select2_multiple");
	 $eventSelect.on("select2:select", function (event) { 
		 event.preventDefault();
	 });
Thanks

**Comment by jmalatia** [comment]
+1
